Nickname Dingbat support

oshaboy opened the pull request:
In the games you can give your Pokémon Emoticons that aren't in Unicode. I implemented them. It's a bit of a hack and I didn't test it a lot so maybe this should be a draft. 

The font is not made by me I downloaded it from Bulbapedia.  

![image](https://github.com/user-attachments/assets/464e42e7-73e1-4c87-a476-0a7eccbaa232)

oshaboy committed:
Pokedings

oshaboy committed:
2

Zarel commented:
Interesting idea. I'm guessing they're in Unicode's reserved space? Probably we'd want a way to input them...

oshaboy commented:
Some of them are over regular codepoints while others are in the PUA

oshaboy commented:
Probably downloading a font from Bulbapedia's internals which I don't know where it came from and putting it into the files verbatim isn't the best idea. And just adding it to the css as a font is too intrusive because it applies to non-PUA codepoints. I should try to find how to make a PUA only version. 

How would a keyboard for dingbats even integrate into the existing UI?

oshaboy commented:
The font might have been taken from the internal files of Pokémon Global Link

oshaboy commented:
The game let you type "halfwidth" dingbats in Western versions of the game and "fullwidth" dingbats in Eastern versions which are internally stored as separate codes to preserve the nicknames. Is this distinction important for Pokémon showdown?

Also Gen V used a different codepoint assignment from the Modern Games. PGLDings on Bulbapedia actually has both of them. 

I would like to clarify that by "intrusive". If you already used certain emoji they will be converted to the PGLDings form upon update. For example, if you have a snowman emoji in your Pokémon nickname will become the two blobs from PGLDings. So either we copy all the dingbats into the PUA or just accept that some nicknames will change.

Zarel commented:
"Keyboard for dingbats" doesn't seem _too_ hard to integrate into the existing UI, just have some buttons for them. The big question is whether this is worth it in the first place, which I have confused feelings about. I'd be curious how PKHeX handles this.

Zarel commented:
Okay, the Bulbapedia article has more details and leans me towards "sure, why not".

Zarel commented:
I guess "PGLDings" comes from some web version of pokemon home that displays them? That seems like a thing we might as well use.

oshaboy commented:
Well for now I will just move all the dingbats to the PUA in fontforge

oshaboy committed:
moved dingbats to PUA

oshaboy committed:
Merge branch 'master' into pokedings

kwsch commented:
> I guess "PGLDings" comes from some web version of pokemon home that displays them?

Was from [PGL](https://bulbapedia.bulbagarden.net/wiki/Pok%C3%A9mon_Global_Link); the font file is what was used by the site to render player-assigned strings from the game.

> I'd be curious how PKHeX handles this.

PKHeX serves the font file for its Nickname/OT text entry fields. It also has implemented the remappings when strings are transferred from one context to the next ([7->8](https://github.com/kwsch/PKHeX/blob/a85f91963052e05a6b1e045f633b0857c00a365d/PKHeX.Core/PKM/Strings/StringConverter8.cs#L191-L230)) as well as the [Font's available characters](https://github.com/kwsch/PKHeX/tree/a85f91963052e05a6b1e045f633b0857c00a365d/PKHeX.Core/PKM/Strings/Font).

For a popup list of the "extra" glyphs not accessible in QWERTY/etc keyboards, PKHeX provides a list of chars that are added to the Trash Byte Editor popup window:
https://github.com/kwsch/PKHeX/blob/a85f91963052e05a6b1e045f633b0857c00a365d/PKHeX.WinForms/Subforms/PKM%20Editors/Text.cs#L227-L334

oshaboy committed:
Revert "moved dingbats to PUA"

This reverts commit fad8d329ec200bde9afa6ea2c2e09b671ea0c12f.

oshaboy commented:
I changed my mind about the PUA assignments.

oshaboy committed:
Implemented Dingbat "Keyboard"

oshaboy committed:
Cleanup

oshaboy committed:
formatting nits

oshaboy commented:
![image](https://github.com/user-attachments/assets/2c06b602-d903-4418-a413-2297e1e4ce0c)

Here's the "Keyboard"

oshaboy committed:
fixed font priority issue.

oshaboy committed:
More formatting.

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Merge branch 'smogon:master' into pokedings

oshaboy committed:
Update client-teambuilder.js

Zarel commented on play.pokemonshowdown.com/style/teambuilder.css:
```diff
@@ -188,6 +188,7 @@
 	padding: 0;
 }
 
+
```
```suggestion
```

oshaboy committed:
Update client-teambuilder.js

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -3680,7 +3687,61 @@
+			this.team = data.team;
 
+			var buf = '<ul name="dingbats" class="popupmenu"><li>';
+			for (var i=0; i<dingbats_halfwidth.length; i++) {
```
```suggestion
			for (var i = 0; i < dingbats_halfwidth.length; i++) {
```

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -3680,7 +3687,61 @@
+			var buf = '<ul name="dingbats" class="popupmenu"><li>';
+			for (var i=0; i<dingbats_halfwidth.length; i++) {
+				if (dingbats_halfwidth[i] == '\n'){
+					buf +='</li><li>'
```
```suggestion
					buf += '</li><li>'
```

oshaboy committed:
Update play.pokemonshowdown.com/js/client-teambuilder.js

Co-authored-by: Guangcong Luo <guangcongluo@gmail.com>

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -3680,7 +3687,61 @@
 
+			var buf = '<ul name="dingbats" class="popupmenu"><li>';
+			for (var i=0; i<dingbats_halfwidth.length; i++) {
+				if (dingbats_halfwidth[i] == '\n'){
```
```suggestion
				if (dingbats_halfwidth[i] == '\n') {
```

oshaboy committed:
Update play.pokemonshowdown.com/js/client-teambuilder.js

Co-authored-by: Guangcong Luo <guangcongluo@gmail.com>

oshaboy committed:
Update teambuilder.css

oshaboy committed:
Merge branch 'pokedings' of https://github.com/oshaboy/pokemon-showdown-client into pokedings

oshaboy committed:
Update play.pokemonshowdown.com/js/client-teambuilder.js

Co-authored-by: Guangcong Luo <guangcongluo@gmail.com>

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -3680,7 +3687,61 @@
+			}
+			buf += '</li><input type="checkbox" name="wideCheckbox"></input>Wide</ul>';
+			this.$el.html(buf);
+			var el=this.$el;
```
We generally would name this variable `$el`, like we do for other jQuery objects.

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Merge branch 'pokedings' of https://github.com/oshaboy/pokemon-showdown-client into pokedings

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -3680,7 +3687,61 @@
+			this.$el.find("input[name=wideCheckbox]").change(
+				function (e) {
+					var width = e.currentTarget.checked?"fullwidth":"halfwidth";
+					$el.find("button[name=insertDingbat]").each(function (_,elem) {
```
```suggestion
					$el.find("button[name=insertDingbat]").each(function (_, elem) {
```

oshaboy committed:
Update play.pokemonshowdown.com/js/client-teambuilder.js

Co-authored-by: Guangcong Luo <guangcongluo@gmail.com>

Zarel commented on play.pokemonshowdown.com/style/client.css:
```diff
@@ -489,6 +493,16 @@ p.or:after {
+	border: transparent;
+	background-color: transparent;
+}
+.dingbat-button:hover{
```
```suggestion
.dingbat-button:hover {
```

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Update play.pokemonshowdown.com/style/client.css

Co-authored-by: Guangcong Luo <guangcongluo@gmail.com>

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Update client-teambuilder.js

oshaboy committed:
Update client-teambuilder.js

Zarel commented:
Okay. I've gotten new information that these don't exist in Gen 8 and later, which makes me less inclined to add them. Although maybe it'd still be neat in past gens.

oshaboy commented:
Gen 8 and 9 use the build in Switch keyboard for nicknames. But it's not like Pokémon Showdown really checks nicknames for in-game legality (For example, It lets you use characters in nicknames that aren't available on the Nintendo Switch keyboard). So I don't see why it shouldn't allow Gen 4 nicknames to be used in Gen 9. 

Though Gen 1 also lets you use characters that aren't in Unicode or Gen 4-7 (Specifically <sup>P</sup><sub>K</sub> and <sup>M</sup><sub>N</sub> which I am simulating using Markdown). The only reason I didn't add them is because PGLDings doesn't support them and I am unaware of a font that does.

oshaboy committed:
Added fallback font formats

oshaboy committed:
Fixed up the CSS and added dingbats to react.

oshaboy committed:
Merge branch 'master' into pokedings

oshaboy committed:
fixed merge conflict

oshaboy committed:
I hate eslint

Zarel committed:
Style

Zarel committed:
JSX style

Zarel committed:
Better way to insert dingbats

(`execCommand` supports Ctrl+Z to undo, and also this approach inserts
at cursor position rather than end.)

Zarel committed:
Redesign

Zarel committed:
Fix wide checkbox in old teambuilder

Zarel commented:
Okay, so, how it works in newclient is great. The oldclient functionality, I'd like the button to only appear after the Nickname box is focused.

oshaboy commented:
I don't see why you think that little smiley face is cluttered. Also wouldn't clicking the button unfocus the nickname box and therefore make the button disappear. Seems like it's more intuitive to keep it always enabled.

oshaboy committed:
Made the design slightly worse for Zarel to fix.

oshaboy committed:
eslint

Zarel committed:
Redesign oldclient button

Zarel committed:
Redesign oldclient menu

Zarel committed:
Merge branch 'master' into pokedings

Zarel commented:
Doesn't actually seem to work reliably. Editing the text in the box immediately make them work correctly, though.

<img width="203" alt="image" src="https://github.com/user-attachments/assets/10fe3177-a54e-4fa6-8b72-df717c9ee523" />

Zarel commented:
`PGLDings, Verdana` instead of `Verdana, PGLDings` seems to fix it... but at what cost?

oshaboy committed:
Merge branch 'master' into pokedings

oshaboy committed:
Fixed Issue with Chrome
